### PR TITLE
Support \d <object> meta command

### LINKED ISSUE #9595: sql: support psql `\d <object>`
pg_catalog does not support all expected columns on the pg_class table. There might be other queries that `\d` executes that are also not supported.

The query is as follows:
```
SELECT c.relchecks,
       c.relkind,
       c.relhasindex,
       c.relhasrules,
       c.relhastriggers,
       c.relrowsecurity,
       c.relforcerowsecurity,
       c.relhasoids,
       FALSE AS relispartition,
                '',
                c.reltablespace,
                CASE
                    WHEN c.reloftype = 0 THEN ''
                    ELSE c.reloftype::pg_catalog.regtype::pg_catalog.text
                END,
                c.relpersistence,
                c.relreplident
FROM pg_catalog.pg_class c
LEFT JOIN pg_catalog.pg_class tc ON (c.reltoastrelid = tc.oid)
WHERE c.oid = $OID;
```

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds the required catalog tables, columns, and functions
required to support the \d <object> meta command. The majority of these
are not applicable to Materialize and filled with sane defaults. The
rest are implemented in a way that makes sense with Materialize.

The output of \d <object> differs slightly from Postgres. The two
biggest differences are that we include primary key indexes while
PostgreSQL doesn't, and our index syntax is different.

Closes #9595

### Motivation

This PR adds a known-desirable feature.

### Testing

- [X] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Improve Postgres catalog compatibility:
    - Add the following columns to the `pg_class` catalog table: `reoftype`, `reltablespace`, `reltoastrelid`, `relhasindex`, `repersistence`, `relchecks`, `relhasrules`, `relhastriggers`, `relrowsecurity`, `relforcerowsecurity`, `relreplident`, `relispartition`, `relhasoids`.
    - Add the following columns to the `pg_index` catalog table: `indisunique`, `indisclustered`, `indisvalid`, `indisreplident`.
    - Add the following columns to the `pg_type` catalog table: `typcollation`.
    - Add the following columns to the `pg_attribute` catalog table: `attgenerated`, `attcollation`.
    - Add the following new catalog tables: `pg_collation`,  `pg_policy`, `pg_inherits`.
    - Modify the `pg_get_constraintdef` function to always return null instead of erroring out.
    - Implement the `pg_get_indexdef` function.
  - Add `USING <index-type>` syntax to index creation. "Arrangement" is the only valid index type.
  These changes enable the `\d <object>` command in the  [psql terminal](/connect/cli) {{% gh 9595 % }}

### CODE REVIEW

File: src/coord/src/catalog/builtin.rs [hunk 1]
```diff
@@ -1557,18 +1557,46 @@ FROM mz_catalog.mz_schemas",
 pub const PG_CLASS: BuiltinView = BuiltinView {
     name: "pg_class",
     schema: PG_CATALOG_SCHEMA,
+    /*
+     * MZ doesn't support typed tables so reloftype is filled with 0
```
Comment (sploiselle) [inline src/coord/src/catalog/builtin.rs:1561]: Can you move these comments inline with the query, i.e. to the appropriate line ? prefixing a line with `--` makes it a comment

File: src/coord/src/catalog/builtin.rs [hunk 2]
```diff
@@ -1594,10 +1622,20 @@ FROM mz_catalog.mz_databases",
 pub const PG_INDEX: BuiltinView = BuiltinView {
     name: "pg_index",
     schema: PG_CATALOG_SCHEMA,
+    /*
```
Comment (sploiselle) [inline src/coord/src/catalog/builtin.rs:1625]: ditto

File: src/expr/src/scalar/func.rs [hunk 3]
```diff
@@ -2593,10 +2600,7 @@ impl BinaryFunc {
             BinaryFunc::LogNumeric => eager!(log_base_numeric),
             BinaryFunc::Power => eager!(power),
             BinaryFunc::PowerNumeric => eager!(power_numeric),
-            BinaryFunc::PgGetConstraintdef => Err(EvalError::Unsupported {
-                feature: "pg_get_constraintdef".to_string(),
-                issue_no: Some(9483),
-            }),
+            BinaryFunc::PgGetConstraintdef => Ok(eager!(pg_get_constraintdef)),
```
Comment (sploiselle) [inline src/expr/src/scalar/func.rs:2603]: This enum should get removed entirely.

File: src/expr/src/scalar/func/impls/oid.rs [hunk 4]
```diff
@@ -70,11 +68,12 @@ sqlfunc!(
 );
 
 sqlfunc!(
-    fn pg_get_constraintdef(_oid: Option<Oid>) -> Result<String, EvalError> {
-        Err(EvalError::Unsupported {
-            feature: "pg_get_constraintdef".to_string(),
-            issue_no: Some(9483),
-        })
+    #[sqlname = "pg_get_constraintdef"]
```
Comment (sploiselle) [inline src/expr/src/scalar/func/impls/oid.rs:71]: This impl should be deleted, as well

**Review by sploiselle — COMMENTED** [review]
To handle this in planning, `pg_get_constraintdef` in `src/sql/src/func.rs` should look like this:

```rust
            "pg_get_constraintdef" => Scalar {
                params!(Oid) => Operation::unary(|_ecx, _oid| {
                    Ok(HirScalarExpr::literal_null(ScalarType::String))
                }), 1387;
                params!(Oid, Bool) => Operation::binary(|_ecx, _oid, _pretty| {
                    Ok(HirScalarExpr::literal_null(ScalarType::String))
                }), 2508;
            },
```

Sorry for the mess!

File: src/expr/src/scalar/func.rs [hunk 5]
```diff
@@ -1130,6 +1130,13 @@ fn power_numeric<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError>
     }
 }
 
+fn pg_get_constraintdef<'a>(_: Datum, _: Datum) -> Datum<'a> {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/expr/src/scalar/func.rs]: no worries, I moved everything to planning. I think it looks a lot cleaner now.

File: src/expr/src/scalar/func.rs (hunk 3 above)
Comment (jkosh44) [reply src/expr/src/scalar/func.rs]: removed

File: src/expr/src/scalar/func/impls/oid.rs (hunk 4 above)
Comment (jkosh44) [reply src/expr/src/scalar/func/impls/oid.rs]: removed

File: src/coord/src/catalog/builtin.rs (hunk 1 above)
Comment (jkosh44) [reply src/coord/src/catalog/builtin.rs]: Moved

File: src/coord/src/catalog/builtin.rs (hunk 2 above)
Comment (jkosh44) [reply src/coord/src/catalog/builtin.rs]: Moved

**Comment by sploiselle** [comment]
@jkosh44 I hesitate to output something we don't take as input, and making a the `USING` argument a nop seems not ideal. I'm curious what the intended use of this is by drivers? If we knew how it gets consumed/what they do with it, I think we'd be able to make a clearer decision.

Barring understanding that, I think we should follow the spirit of the command, rather than trying to match its output verbatim. 

My output looks a little different than what you mention


```
sean=# create table xy (a int);
sean=# create index foo_a on xy(a);
CREATE INDEX
sean=# \d foo_a;
         Index "public.foo_a"
 Column |  Type   | Key? | Definition 
--------+---------+------+------------
 a      | integer | yes  | a
btree, for table "public.xy"
```

Am I misreading your comment, or is your output really that distinct-looking from mine?

**Comment by jkosh44** [comment]
@sploiselle Sorry, you're not misreading my comment, I just wasn't very clear. I was talking about the output when using `\d` on a table so 
```
postgres=# create table xy (a int);
CREATE TABLE
postgres=# create index foo_a on xy(a);
CREATE INDEX
postgres=# \d xy
                 Table "public.xy"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 a      | integer |           |          | 
Indexes:
    "foo_a" btree (a)
```

When I run it against the index, I get the same output as you
```
postgres=# \d foo_a
         Index "public.foo_a"
 Column |  Type   | Key? | Definition 
--------+---------+------+------------
 a      | integer | yes  | a
btree, for table "public.xy"
```

**Comment by jkosh44** [comment]
@sploiselle I also agree with you on not outputing something we don't take as input. From some very brief searching on GitHub it seems like at least some people use the output of `pg_getindexdef` to recreate indexes which would not work if we outputed invalid input.

On the other hand I'm also worried that something may try and parse the output of `\d <table>` and fail because the indexes are not formatted properly.

**Comment by sploiselle** [comment]
@jkosh44 I see! Asked about this in [Slack](https://materializeinc.slack.com/archives/CMHDK0DK8/p1646268234767799) and determined our index structures are distinct and there's no `USING` clause that would really be appropriate (and really kind of useless given that we're very unlikely to offer an arrangement with some radically different structure).

I'll take a closer look tomorrow.

**Comment by jkosh44** [comment]
> @jkosh44 I see! Asked about this in [Slack](https://materializeinc.slack.com/archives/CMHDK0DK8/p1646268234767799) and determined our index structures are distinct and there's no `USING` clause that would really be appropriate (and really kind of useless given that we're very unlikely to offer an arrangement with some radically different structure).
> 
> I'll take a closer look tomorrow.

I took a look at the conversation. We don't necessarily need to draw a parallel between our index types and an existing Postgres index type. I was thinking that we'd come up with our own name like "arrangement". In fact Postgres allows you to define your own index types and index names (https://www.postgresql.org/docs/current/sql-createindex.html)
> PostgreSQL provides the index methods B-tree, hash, GiST, SP-GiST, GIN, and BRIN. Users can also define their own index methods, but that is fairly complicated.

So I was thinking something like `create index foo_a on xy using arrangement (a);` would be valid syntax and be parsed correctly, but it would have the exact some affect as `create index foo_a on xy(a);`. Similar to how Postgres allows you to specify an index type, but will default to `btree` if it's omitted. 

That way `pg_get_indexdef` would output `CREATE INDEX foo_a ON xy USING arrangement (a)` and `\d xy` would output
```
                 Table "public.xy"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 a      | integer |           |          | NULL
Indexes:
    "foo_a" arrangement (a)
    "xy_primary_idx" arrangement (a)
```

**Comment by sploiselle** [comment]
@benesch @frankmcsherry How would you all feel about labeling our "indexing structure" as "arrangement"? This would e.g. parallel `btree` in PG as the default structure of indexes. Joe's comment above this provides more detail.

**Comment by frankmcsherry** [comment]
Seems good to me. Do I understand rightly that the question is just "is the term `arrangement` the most appropriate term for 'the index implementation strategy'"?

**Comment by benesch** [comment]
Oooh yes!

**Comment by sploiselle** [comment]
@jkosh44 Yeah--I think the consensus is to move that into this PR so we can make the output into the shape we'd like

**Comment by jkosh44** [comment]
@sploiselle Cool, just pushed the changes included in that patch.

**Review by sploiselle — COMMENTED** [review]
This is great!

Two additional asks:
- Can we add a test for this _a la_ https://github.com/MaterializeInc/materialize/pull/9346/files#diff-7e0bb78a74e93598a5b4d53e78dc016350ed9f9061b5766925af26f639b0ae72
- We should remove the experimental mode requirement for `array_cat`, which this change requires. The thinking behind this is that I am a stickler for avoiding array support except for as much as it's required for PG catalog compatibility, and this is in the catalog so qualifies `array_cat` to no longer require experimental mode.

File: doc/user/data/sql_funcs.yml [hunk 6]
```diff
@@ -685,7 +685,14 @@
   - signature: 'pg_column_size(expr: any) -> int'
     description: Returns the number of bytes used to store any individual data value.
   - signature: 'pg_get_constraintdef(oid: oid [, pretty: bool]) -> text'
-    description: Returns the constraint definition for the given `oid`. Currently always returns an error.
+    description: >-
+      Returns the constraint definition for the given `oid`. Currently always
+      returns NULL since constraints aren't supported.
+  - signature: ' pg_get_indexdef(index: oid [,column integer, pretty: bool]) -> text'
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply doc/user/data/sql_funcs.yml]: Fixed both

File: src/coord/src/catalog/builtin.rs [hunk 7]
```diff
@@ -1561,14 +1561,40 @@ pub const PG_CLASS: BuiltinView = BuiltinView {
...
+    -- TODO(jkosh44): update this column when issue is resolved.
+    'p'::pg_catalog.\"char\" AS relpersistence,
     CASE
         WHEN mz_objects.type = 'table' THEN 'r'
         WHEN mz_objects.type = 'source' THEN 'r'
         WHEN mz_objects.type = 'index' THEN 'i'
         WHEN mz_objects.type = 'view' THEN 'v'
-    END relkind
+    END relkind,
+    -- MZ doesn't support constraints so relchecks is filled with 0
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/coord/src/catalog/builtin.rs]: Fixed

**Comment by jkosh44** [comment]
@sploiselle I added a test to the same file and removed experimental mode from `array_cat`

**Review by sploiselle — APPROVED** [review]

**Pull request merged**